[RVV] Optimize Generic RVV Matmul codegen

bhbruce opened the pull request:
### Background
RVV matmul always uses register tile size [m, n] = [8, 32]. It's ok for VLEN=512. It can generate outer product GEMM with LMUL=2. The vector register utilization rate is 56%(18/32). However, it uses LMUL=1 for VLEN=1024 and the vector register utilization rate is 28%. If we use VLEN = 128. The codegen uses LMUL=8 and vector registers are not enough. It introduces register spill and causes performance drop.
### Optimization
Two optimizations are included in this PR to improve RVV matmul code generation.

1. Register tiling optimization: We introduce VLEN-aware register tiling. The register tile size is changed from [m, n, k] = [8, 32, 1] to [7, vlmax_lmul_4, 1]. This can not only fix the problem of generating inefficient code for `VLEN!=512` but also increase the vector register utilization rate to 100%. Based on experiments on SiFive platforms, the performance improvement is around 25%~50%.

2. Cache tiling optimization: 
  - After introducing VLEN-aware register tiling, the original cache tiling code may break since `maxTileSize` is set to 64 by default. To solve the problem, we guarantee `maxTileSize >= minTileSize`, which implies that cache tile size >= register tile size.
  -   Introducing option `iree-llvmcpu-aggressive-distribution` makes cache tile size become twice of register tile size if `maxTileSize < minTileSize`. It reduces cache miss rate and leads to performance improvement. Performance improvement is 3~5% in general.

In summary, my experiment on the SiFive VLEN 1024 platform shows x5.4 speedup for Bert-Large after applying the patches.

bhbruce committed:
Optimize RVV matmul gemmv codegen: use m8 instead of m4

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Add hasAnyVFeature() to detect RVV vector feature

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Add RVV 7 x m4 matmul lowering_config & tile_and_fuse test

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Decide matmul output tile size by VLEN and set to [7, m4].

For RISC-V path, matmul output tile size is set to [m,n,k]=[8,32,1] by default.
It is efficient for vl = 512. It generates code using LMUL=2 and 18 vector registers.
It makes the vl != 512 cases generate inefficient code.
For vl = 1024, it will generate LMUL=1 code which only use 8+1 vec registers.
For vl = 256, it will generate LMUL=4 code which introduce vector register spill.

This commit detects if RISC-V Vector is supported, then calling `getMatmulRISCVVectorSizes`.
This function aims to fully utilize vector registers, so it generates tile size [7, vlmax_lmul4, 1].
Using lmul4 can fully utilize all 32 vector registers.

For now, it only applies to `nonWideningLinalgElementType`.
It uses `native_vector_size` and data_type to determine `vl for LMUL=4`.

Comparing tile setting [8, m2] vs [7, m4] on XNNPACK using SiFive platforms shows performance
improvement around 25%~50%.

Example:
`native_vector_size` is 256 for vl=1024 & default LMUL=2.
  The output tile size:
  m = 7
  n = 256 * 8 * 2 / 32 = 128 => equals to vlmax (SEW=32 & LMUL=4)
  k = 1

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Avoid determining tile size w/ VLEN for matmul_transpose_b

Use default tile size for matmul_transpose_b & batch_matmul_transpose_b
to avoid performance drop when using getMatmulVectorSizesUsingFullVectorHeuristics.

If tile size is set to [m, n] = [8, 32], the output asm uses vector slide and fmacc instuctions
in the inner most loop. This results in acceptable performance, though not optimal.

However, if we increase the tile size along the N dimension, the output asm includes
numerous scalar load and store instructions within the innermost loop, which degrades performance.

Therefore, we use the default tile size for matmul_transpose_b.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Add iree-llvmcpu-aggressive-distribution option

This option enlarges maxTileSizes to twice of minTileSizes if maxTileSizes > minTileSizes.
To make cache tile size become twice to register tile size.

Assume VLEN=1024, so register tile size [m, n] = [7, 128].
By default, maxTileSizes[n-dim] = 64.
This option makes maxTileSizes change to 256, i.e twice of n-dim register tile size.
In other word, n-dim cache tiling is guaranteed and reduces cache miss rate.

The performance improves 3~5% in general cases.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Add RVV 1 x m4 GEMV lowering_config test

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Make matmul cache_tile_size >= register_tile_size in n-dim for RVV

Default maximum cache_tile_size is clDefaultDistTileSize(64).
If RVV is enabled and vl = 1024, the register_tile_size is [m, n, k]=[7, 128, 1].

In this case, minTileSizes[n-dim] equals to 128(register_tile_size in n-dim),
it makes maxTileSizes < minTileSizes in n-dim.

This commit guarantees maxTileSizes >= minTileSizes and it imply that
cache_tile_size >= register_tile_size in n-dim.

Note: Only apply matmul cache_tile_size modification for RVV now to workaround
test `select_x86_64_lowering_strategy.mlir.test` fail.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -61,6 +61,14 @@ static llvm::cl::opt<int>
+        "It is only applied for linalg contraction ops now. "
+        "If distConfig.minTileSizes[i] >= distConfig.maxTileSizes[i], "
+        "set distConfig.maxTileSizes[i] to 2 * distConfig.minTileSizes[i]."),
+    llvm::cl::init(false));
```
Is this for risc-v only? Can we add `-riscv-` to the flag for now? I don't see a plan that we'll use it for all the targets at this moment.

Also, could you move it to line 108 (i.e., below `clDisableArmSMETiling`), so we can group all the flags related to CPU targets in the same place?

bhbruce replied:
Yes, it's for RVV only. I modify it to `iree-llvmcpu-riscv-aggressive-distribution` and move it to line 108.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+  } else if (elementType->isF64()) {
+    elementSize = 64;
+  } else {
+    // ToDo: support int data type
```
style nit: TODO

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+                          SmallVectorImpl<bool> &scalableSizeFlags) {
+  if (sizes.empty())
+    getDefaultMatmulVectorSizes(op, vectorSize, sizes, scalableSizeFlags);
+  // Todo: support widening matmul.
```
style nit: TODO

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+  // Todo: support widening matmul.
+  // Determines n dimension tile size with VLEN for
+  // nonWideningLinalgElementType.
+  auto elementType = nonWideningLinalgElementType(op);
```
Can you spell out the type? It is `FailureOr<Type>`.

bhbruce replied:
Thank you for catching it. Let me modify it.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+  auto lhsShape = lhsType.getShape();
+  bool has_batch = isa<linalg::BatchMatmulOp, linalg::BatchMatmulTransposeBOp>(
+      op.getOperation());
+  char idx = (has_batch) ? 1 : 0;
```
nit: use `int`. Should we use it in all those `sizes` access (e.g., should below all become `sizes[baseIdx+0/1/2/3/4]`? If so, I think we can rename it to `baseIdx`.

bhbruce replied:
This `idx` aims to point to `m dimension index` in `lhsShape`.
How about changing to `mDimIdx`?

hanhanW replied:
I see. In this case, I think what you're looking for is [inferContractionDims](https://github.com/llvm/llvm-project/blob/15d1560ea4047a2b4b14c826767089f538ddda70/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.h#L50-L65). This provides a more flexible way to not just look at named ops. You can get the batch/m/n/k dimensions from it.

Furthermore, you can bail out if one of the dimension is greater than one.

EDIT: This is an optional review comment. I don't know your end goal, but I'd like to share that there is an utility for inferring these dimensions.

bhbruce replied:
Thank you for providing this approach. However, I am confused in why each dimension has size two.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+    return;
+  }
+  Value lhs = op.getDpsInputs()[0];
+  ShapedType lhsType = cast<ShapedType>(lhs.getType());
```
style nit: we follow mixed LLVM and Google coding style in IREE. We should use `auto` here because `cast` already spells the type.

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

bhbruce replied:
ok.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1297,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+  // Use 7 x lmul4 to fully utilize vector registers.
+  sizes[0] = 7;
+  // Calculate tile size for the main vector dimension (N).
+  constexpr int64_t byteSizeInBits = 8;
```
style nit

https://google.github.io/styleguide/cppguide.html#Constant_Names

```suggestion
  constexpr int64_t kByteSizeInBits = 8;
```

bhbruce replied:
solved.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1354,6 +1420,18 @@ getMatmulVectorSizes(mlir::FunctionOpInterface entryPointFn,
+    bool isBMTB = isa<linalg::BatchMatmulTransposeBOp>(op);
+    // Use default tile size for matmul_transpose_b &
+    // batch_matmul_transpose_b to avoid performance drop.
+    if (!isMTB && !isBMTB) {
```
nit: you can collapse it to one `isa` in the if condition.

```suggestion
    // Use default tile size for matmul_transpose_b &
    // batch_matmul_transpose_b to avoid performance drop.
    if (!isa<linalg::MatmulTransposeBOp, linalg::BatchMatmulTransposeBOp>(op)) {
```

bhbruce replied:
Applied suggestion.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile_and_fuse_riscv.mlir:
What is the purpose of adding this test? IMO, it really looks like a test with transform spec, although you're using C++ passes. It is not really maintainable in the IREE main repo because it is testing a combination of passes. The lowering_config is preset, and the TileAndFuse tests are covered in the other file.

If we want to do the pipeline test, are you able to add a test in [pipeline_tests.mlir](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir)? Or create a `pipeline_aggressive_distribution_riscv_tests.mlir` and add the test to the new file? (Note, you'll need to use `builtin.module(iree-llvmcpu-select-lowering-strategy, func.func(iree-llvmcpu-lower-executable-target))` but not these arbitrary passes.)

The main difference is that the pipeline tests run the full pipeline which includes strategy selection and the lowering. This gives us a better picture about what we're expecting in the test.

bhbruce replied:
> What is the purpose of adding this test?

I would like to ensure that the compiler can help generate the desired for-loop struct if using register tile size [7, 64] and cache tile size [64,128].

I agree with you. I should create the other file to run the whole pipeline test. Thank you for providing the details.

hanhanW replied:
E.g., here is a contraction op but the number of M dimensions is more than 1.

```mlir
util.func public @multi_m_dim_generic(%arg0 : tensor<64x4x128xf32>, %arg1 : tensor<128x512xf32>,
    %arg2 : tensor<64x4x512xf32>) -> tensor<64x4x512xf32> {
    %4 = linalg.generic {
        indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>,
                         affine_map<(d0, d1, d2, d3) -> (d2, d1)>,
                         affine_map<(d0, d1, d2, d3) -> (d0, d3, d1)>],
        iterator_types = ["parallel", "parallel", "parallel", "reduction"]}
        ins(%arg0, %arg1 : tensor<64x4x128xf32>, tensor<128x512xf32>) outs(%arg2 : tensor<64x4x512xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %5 = arith.mulf %in, %in_0 : f32
      %6 = arith.addf %5, %out : f32
      linalg.yield %6 : f32
    } -> tensor<64x4x512xf32>
  util.return %4 : tensor<64x4x512xf32>
}
```

bhbruce replied:
Do we face this kind of MLIR in simple pass routing?
I supposed that multi_m & multi_n only happens when enabling mmt4d.

hanhanW replied:
I don't have handy IR/examples now. However, it is not mmt4d specific. We've seen that in some models. One of design points of the data-tiling (e.g., mmt4d things) is that we can narrow down the variants for GEMM kernels.

bhbruce replied:
OK, thank you for clarifying. I'll keep in mind for this.

hanhanW reviewed (changes requested):
Thanks for the contribution! I left some comments, please take a look.

bhbruce committed:
Change aggressive option name to include riscv

Change From `iree-llvmcpu-aggressive-distribution` to `iree-llvmcpu-riscv-aggressive-distribution`.
Also, fix some coding style problems.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Remove tile_and_fuse_riscv test

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Add riscv-aggressive-distribution pipeline test

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1289,6 +1298,64 @@ static void getMatmulVectorSizesUsingFullVectorHeuristics(
+  }
+  Value lhs = op.getDpsInputs()[0];
+  auto lhsType = cast<ShapedType>(lhs.getType());
+  auto lhsShape = lhsType.getShape();
```
I did not notice that what you need is just the shape. How about using some linalg interface method to get it? You can do something like `ArrayRef<int64_t> lhsShape = linalgOp.getShape(linalgOp.getDpsInputOperand(0));`. Then we don't need the `lhs` and `lhsType` variables.

https://github.com/llvm/llvm-project/blob/1f2509993e6e0717b547b5214b06550af4f3008f/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L417-L439

bhbruce replied:
I'm new to MLIR. Thank you for the info.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy_with_aggressive_distribution.mlir:
I think we can collapse this file into the other one. What you're looking for is adding another `// RUN...` with the flag. E.g.,

https://github.com/iree-org/iree/blob/c651ba9257315af96e75a371537f3cf8ab080cb0/compiler/src/iree/compiler/Codegen/Common/test/decompose_pack_unpack_ops.mlir#L1-L2

This way reduces the dup tests in different files.

bhbruce replied:
OK. I'll collapse it to `compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir`.

hanhanW replied:
Oops, sorry for not being clear enough. What I meant is collapsing the `select_riscv_lowering_strategy_*.mlir` to the `select_riscv_lowering_strategy.mlir` file, not the pipeline_test one.

The original `pipeline_riscv_aggressive_distribution_tests.mlir` looks good to me, can you recover it?

bhbruce replied:
oh... My bad... I misunderstood it. Let me revert and merge `select_riscv_lowering_strategy_*.mlir`.

bhbruce replied:
Solved!

hanhanW reviewed (commented):
Overall looks good to me, just few nits.

bhbruce committed:
Merge pipeline_riscv_aggressive_distribution_tests to pipeline_tests

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Use linalg::inferContractionDims to infer m-dim index

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Revert "Merge pipeline_riscv_aggressive_distribution_tests to pipeline_tests"

This reverts commit d0e0c8cc8e10d9433c6354e38ba2ae92c91a355b.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Merge select_riscv_lowering_strategy*.mlir test

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir:
```diff
@@ -30,6 +31,105 @@ func.func @matmul_riscv() attributes {hal.executable.target = #executable_target
+
+#executable_target_embedded_elf_riscv_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-riscv_64", {cpu_features = "+m,+a,+f,+d,+zvl1024b,+v", data_layout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128", native_vector_size = 256 : index, target_triple = "riscv64-unknown-unknown-eabi-elf"}>
+builtin.module {
+  func.func @matmul_riscv_vl1024() attributes {hal.executable.target = #executable_target_embedded_elf_riscv_64_} {
```
What is the difference between this `matmul_riscv_vl1024` test and the below `matmul_riscv_vl1024` test? Can we collapse them into one? (In this case, you can move all your `CHECK-AGGRESSIVE`s right after the regular `CHECK`s). E.g.,

```
module {....}
// CHECK: ...
// CHECK: ...
// CHECK: ...
...

// CHECK-AGGRESSIVE: ...
// CHECK-AGGRESSIVE: ...
// CHECK-AGGRESSIVE: ...
...

bhbruce replied:
The matmul problem sizes were slightly different between them. 
I merged two MLIR files into one to avoid confusion.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_riscv_lowering_strategy.mlir:
```diff
@@ -30,6 +31,105 @@ func.func @matmul_riscv() attributes {hal.executable.target = #executable_target
+
+#executable_target_embedded_elf_riscv_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-riscv_64", {cpu_features = "+m,+a,+f,+d,+zvl512b,+v", data_layout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128", native_vector_size = 128 : index, target_triple = "riscv64-unknown-unknown-eabi-elf"}>
+builtin.module {
+  func.func @gemv_riscv_vl512() attributes {hal.executable.target = #executable_target_embedded_elf_riscv_64_} {
```
Same question, I can't distinguish the difference between this and the above one based on the name. What is the difference?

bhbruce replied:
This is different in problem size. LHS is 1 by 512, so it is GEMMV. 
Other tests are GEMM test.

hanhanW replied:
I see, can you put the "gemmv" and "gemm" words to the function name in this case?

bhbruce replied:
Ok. I also fix the ci error in previous commits. Let's wait for the result.

hanhanW reviewed (commented):
Wh

EDIT: I did not mean to submit this comment, please ignore this one.

bhbruce committed:
Reduce test cases in select_riscv_lowering_strategy.mlir

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

hanhanW reviewed (commented):
LGTM with the final nit: https://github.com/iree-org/iree/pull/18986#discussion_r1847163822

Thanks for addressing my comments!

bhbruce committed:
Merge branch 'main' into rvv-matmul

bhbruce committed:
Add GEMM or GEMV in select_riscv_lowering_strategy.mlir test name

Also, fix the ci error because lacking of `pipeline = ` before `CPUDoubleTilingExpert`.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce committed:
Fix error by checking for m.size()

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>
